modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl.java

```diff
@@ -522,7 +525,11 @@ protected boolean compareAndSetGlobalState(final int expVal, final int newVal,
                                 GridCacheLockState val = lockView.get(key);
 
                                 if (val == null)
-                                    throw new IgniteCheckedException("Failed to find reentrant lock with given name: " + name);
+                                    if (reCreate) {
```

Comment:
What happens if failoverSafe == false? I'm thinking, if user opted for semaphore that explicitly throws exceptions if something goes wrong, do we need to inform him that new state has been created?
Also I think we should write to log if new state is created.